File: leet_code/python/20.ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        dic = {')':'(',']':'[','}':'{' }
        tmp = []
        for item in s:
            if item in dic.values():
                tmp.append(item)
            elif item in dic.keys():
                if len(tmp) == 0 or dic[item] != tmp[-1]:
                    return False
                else:
                    tmp.pop(-1)
        return True if len(tmp) == 0 else False

if __name__ == "__main__":
    s = Solution()
    print(s.isValid(']'))

File: cpp/20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */
#include <stack>
#include <string>
#include <unordered_map>

using namespace std;

class Solution
{
public:
  bool isValid(string s)
  {
    stack<char> record;
    for (auto c : s) {
      switch (c) {
        case '(':
        case '[':
        case '{':
          record.push(c);
          break;
        case ')':
          if (record.empty() || record.top() != '(')
            return false;
          else
            record.pop();
          break;
        case ']':
          if (record.empty() || record.top() != '[')
            return false;
          else
            record.pop();
          break;
        case '}':
          if (record.empty() || record.top() != '{')
            return false;
          else
            record.pop();
          break;
        default:
          break;
      }
    }
    return record.empty();
  }
};
